### ISSUE TITLE
Add an interceptor to every incoming request

### ISSUE THREAD

**Issue by giangpham96** [zalando/skipper#741]
I want to implement a gateway, in which for every request coming, it calls another service to verify the expiration time of a token. If the token is still valid, then the request will be processed.

Is there any way that I can implement such an interceptor/middleware like that in skipper?

**Comment by giangpham96** [comment]
@szuecs oh, sorry for bothering. That docs you provided probably help

**Comment by szuecs** [comment]
If it does not please reopen the issue and we can discuss how to help you. The feature I have in mind could be done very simple, so it will not take a lot of time to do.

**Comment by giangpham96** [comment]
Hi @szuecs 
Seems like another problem arises
After having the OAuth token introspection in place, I still need to call another service before process the request. Is there any way that I can call a service in skipper? Honestly, I'm just a beginner, so I really don't know if it is possible

**Comment by giangpham96** [comment]
For example, I get the `sub` field from `tokenintrospection`. And after that, I have another request (A) call that consumes the `sub` field as a query parameter. Request (A) returns a field that is needed in the header of all requests behind the gateway.

**Comment by szuecs** [comment]
Let me rephrase it, to make sure I understand you correctly:

You want to set the value of the  `sub` field into a chosen header field to the backend call.

If this is the case  and the rule is only to copy the value into a chosen fixed header, we can easily build a filter that you do not need to have a 3rd service. 
For example: `copyFromTokenintrospectionStateToHeader("sub", "X-Foo-Bar")` would copy the value of `sub` to the Header `X-Foo-Bar`.
We already have the full data from the tokenintrospection response in out "stateBag" and can use this to extract data and put it into a field with a filter, which is probably done in less than 20 lines of code + copy some boilerplate.
If you need to have some **decision** in an external service, this would be another filter todo, which is also not complicated.

**Comment by giangpham96** [comment]
Ah, you're half right.
What I want is that after I get the sub field, I want to make a request, let call it `getRequiredHeader`, and this `getRequiredHeader` wants the sub field to be its query parameter (ie. ?q=sub). The `getRequiredHeader` response contains a field called `requiredHeader`.
After that, I want to add `requiredHeader` to X-Foo-Bar header of all endpoints behind the gateway.

Request coming in -> OAuth introspect -> getRequiredHeader -> Add the header to the request

**Comment by giangpham96** [comment]
But `copyFromTokenintrospectionStateToHeader("sub", "X-Foo-Bar") ` can also be used in another use case :D

**Comment by szuecs** [comment]
It sounds a bit like an OpenIDConnect flow, with a copy from some data to put into a header to send it to the backend. 
Do you want to have an OpenID Connect UI flow like authcode flow?
I have an open PR, which I worked on before and just pushed the current state https://github.com/zalando/skipper/pull/743.

**Comment by giangpham96** [comment]
Ah, never mind, I manage to write my own custom filter. It was a good learning experience!!!

**Comment by szuecs** [comment]
@giangpham96 feel free to open also a PR if you think it makes sense to share this.

**Comment by giangpham96** [comment]
@szuecs I would if I can find a way to generalize the solution.
For now, what I did is very unambiguous, which means it can be applied for our use case 😄

**Comment by szuecs** [comment]
@giangpham96 You can probably generalize it if you split it into multiple filters.
Just an example:

1. https://github.com/zalando/skipper/issues/744
2. copyQueryToHeader(string, string)
